core/server/models/base/index.js

```diff
@@ -763,7 +775,12 @@ ghostBookshelf.Model = ghostBookshelf.Model.extend({
         const id = options.id;
         const model = this.forge({id: id});
 
-        data = this.filterData(data);
+        data = this.filterData(data, {
+            methodName: 'edit',
+            context: options.context
+        });
+
+        data.id = options.id;
```

Comment:
Not sure if this is okay - should I force an error here if data.id is set and is different to options.id? Or is this okay?

Reply:
Why is this line needed now?

Reply:
This is to stop the setting of an id via `edit`, like with a `PUT`.
Though i just realised this should check the context.internal flag too

Reply:
🤔 Shouldn't this be handled via

> return _.without(basePermittedAttributes, 'id');

if action is `edit`.

Reply:
Initially I thought yes, but [kevin mentioned](https://github.com/TryGhost/Ghost/pull/9857#discussion_r215881074) that it should be okay to include the `id` when making a `PUT` or `PATCH` just not to modify it.

Reply:
Looking back at kevins comment he suggests it would be good UX to return an error

Reply:
Typical workflow is a fetch an existing resource from the API, change some values on the returned object then push it back to the API via PUT/PATCH. If we disallow `id` from the PUT and PATCH requests then we force every API consumer to know about this and to filter the property from the request.

I would be in favour of throwing an error if the query param `id` and the submitted resource `id` doesn't match because it's a better UX for the API consumer - something is clearly wrong in their code if they are submitting data to one resource's URL with a different resource's data. Silently applying one resources to a different resources data would be very unexpected.

POST is a special case because it's a new resource, there should never be an `id` property because it's an unknown value until the server has actually saved it, good to error early in that case.

Reply:
> If we disallow id from the PUT and PATCH requests then we force every API consumer to know about this and to filter the property from the request.

We would not disallow and throw error, we would just omit the id property.

> I would be in favour of throwing an error if the query param id and the submitted resource id doesn't match because it's a better UX for the API consumer 

Yeah agree. If options.id !=== data.id that should throw an error.

Reply:
Cool, thanks both :relaxed: